### PR TITLE
Implement a visitor like class for KIR, move passes to it.

### PR DESCRIPTION
**Pull request by csarofeen**
Implements new KirVisitor which helps reduce replicated code in passes. Will extend this in future PRs to have insertion/replacement of Exprs.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.cpp [hunk 1]
```diff
@@ -478,6 +478,42 @@ void OptOutDispatch::handle(GridBroadcast* stmt) {
 void OptOutDispatch::handle(GridWelford* stmt) {
   unhandled(stmt);
 }
+
+std::vector<Expr*> KirVisitor::handle(const std::vector<Expr*>& exprs) {
+  exprs_ = std::vector<Expr*>(exprs);
+  for (auto expr : exprs_) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.cpp]: Isn't `exprs_` potentially modified during a traversal? Shouldn't `exprs` be used here instead?

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 2]
```diff
@@ -590,7 +575,7 @@ class AllocationInserter : public kir::OptOutDispatch {
       // loops
       if (alloc.buffer->memoryType() == MemoryType::Shared &&
           !kir::ExpressionEvaluator::isConst(alloc.alloc_expr->size())) {
-        loop_nests_.insert(loop_nests_.begin(), alloc.alloc_expr);
+        exprs_.insert(exprs_.begin(), alloc.alloc_expr);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp:580]: Here, `exprs_` is modified, so using it with the for loop in `KirVisitor` seems invalid.

**Review by naoyam — COMMENTED** [review]
Overall, very nice cleanup! Just wanted to check one thing in `kir::Visitor::handle`.

File: torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h [hunk 3]
```diff
@@ -136,6 +136,33 @@ class TORCH_CUDA_CU_API OptInDispatch : public OptOutDispatch {
...
+// Includes visiting through scopes like IfThenElse and ForLoop, and tracks them
+// in scopes_ and for_loops_.
+//
+// Makes a copy of exprs at exprs_ which could be used to modify and return.
+//
+// When traversing through ITE/FLs it will use a copy
+// of the provided expressions to make it safe to insert/delete nodes.
+//
+// Provides a simple base class to inherit from for typical kir passes
+class KirVisitor : public OptOutDispatch {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h:150]: nit: Do we need `Kir` in the name here? It's in the `kir` namespace. Maybe just `IrVisitor`?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/kernel_ir_dispatch.h:150]: Yeah, I realized that in the next PR can rename it in a follow up.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**